_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+
+```java
+public AppUser(UserIdentity user) {
+    this.handle = user.getId();
+    this.username = user.getName();
+    this.displayName = user.getDisplayName();
+}
+
+public UserIdentity toUserIdentity() {
+    return UserIdentity.builder()
+        .name(getUsername())
+        .displayName(getDisplayName())
+        .id(getHandle())
+        .build();
+}
+```
+
+Create a Spring Data [`CrudRepository`](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/CrudRepository.html) to manage the `AppUser` objects:
+
+```java
+import com.yubico.webauthn.data.ByteArray;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository extends CrudRepository<AppUser, Long> {
+    AppUser findByUsername(String name);
+    AppUser findByHandle(ByteArray handle);
+}
+```
+
+### Credential data
+
+The server also stores information about the user's credentials. There is a standard `id` and a `name` field used to identify the authenticator on the server side.
+
+To authenticate a credential, the server constructs an `AssertionRequest` Java object that contains the server-side information about any credential. The `AssertionRequest` is converted into JSON and sent to the browser, which it turns into `PublicKeyCredentialRequestOptions` during the WebAuthn authentication ceremony. The required fields are a `credentialId`, a random byte array created by the authenticator that identifies the scope of a credential. A `publicKey` is also required, which is another array of bytes. This is generated by the authenticator's cryptographic algorithm and is used to check the authenticity of a credential signature provided during authentication.
+
+```java
+import com.yubico.webauthn.RegistrationResult;
+import com.yubico.webauthn.data.AttestedCredentialData;
+import com.yubico.webauthn.data.AuthenticatorAttestationResponse;
+import com.yubico.webauthn.data.ByteArray;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Lob;
+import javax.persistence.ManyToOne;
+import java.util.Optional;
+
+@Entity
+@Getter
+@NoArgsConstructor
+public class Authenticator {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+
+    @Column
+    private String name;
+
+    @Lob
+    @Column(nullable = false)
+    private ByteArray credentialId;
+
+    @Lob
+    @Column(nullable = false)
+    private ByteArray publicKey;
+
+    @ManyToOne
+    private AppUser user;
```

Comment:
This original intent was to break up the block with some context.

But I'll move that to a comment block or a notes at the bottom to avoid confusion